// @peculiar/asn1-ecc/build/es2015/object_identifiers.js
export const id_ecPublicKey = "1.2.840.10045.2.1";
export const id_ecDH = "1.3.132.1.12";
export const id_ecMQV = "1.3.132.1.13";
export const id_ecdsaWithSHA1 = "1.2.840.10045.4.1";
export const id_ecdsaWithSHA224 = "1.2.840.10045.4.3.1";
export const id_ecdsaWithSHA256 = "1.2.840.10045.4.3.2";
export const id_ecdsaWithSHA384 = "1.2.840.10045.4.3.3";
export const id_ecdsaWithSHA512 = "1.2.840.10045.4.3.4";
export const id_secp192r1 = "1.2.840.10045.3.1.1";
export const id_sect163k1 = "1.3.132.0.1";
export const id_sect163r2 = "1.3.132.0.15";
export const id_secp224r1 = "1.3.132.0.33";
export const id_sect233k1 = "1.3.132.0.26";
export const id_sect233r1 = "1.3.132.0.27";
export const id_secp256r1 = "1.2.840.10045.3.1.7";
export const id_sect283k1 = "1.3.132.0.16";
export const id_sect283r1 = "1.3.132.0.17";
export const id_secp384r1 = "1.3.132.0.34";
export const id_sect409k1 = "1.3.132.0.36";
export const id_sect409r1 = "1.3.132.0.37";
export const id_secp521r1 = "1.3.132.0.35";
export const id_sect571k1 = "1.3.132.0.38";
export const id_sect571r1 = "1.3.132.0.39";
